fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/UnnestedRecordTypeBuilder.java

```diff
@@ -0,0 +1,290 @@
+            return child;
+        }
+
+        @Nonnull
+        public KeyExpression getNestingExpression() {
+            return nestingExpression;
+        }
+
+        public UnnestedRecordType.Nesting build(List<SyntheticRecordType.Constituent> constituents) {
+            SyntheticRecordType.Constituent parentConstituent = constituents.stream()
+                    .filter(c -> c.getName().equals(parent))
+                    .findFirst()
+                    .orElseThrow(() -> new MetaDataException("unable to find parent constituent")
+                            .addLogInfo(LogMessageKeys.EXPECTED, parent)
+                            .addLogInfo(LogMessageKeys.RECORD_TYPE, getName()));
+            SyntheticRecordType.Constituent childConstituent = constituents.stream()
+                    .filter(c -> c.getName().equals(child))
+                    .findFirst()
+                    .orElseThrow(() -> new MetaDataException("unable to find child constituent")
+                            .addLogInfo(LogMessageKeys.EXPECTED, child)
+                            .addLogInfo(LogMessageKeys.RECORD_TYPE, getName()));
+            return new UnnestedRecordType.Nesting(parentConstituent, childConstituent, nestingExpression);
+        }
+    }
+
+    /**
+     * Internal constructor for creating {@link UnnestedRecordTypeBuilder}s. To create a type and add it to a
+     * meta-data object, use {@link RecordMetaDataBuilder#addUnnestedRecordType(String, String)}.
+     *
+     * @param name name of the type
+     * @param recordTypeKey record type key for the type
+     * @param metaDataBuilder meta-data object this will be defined on
+     * @param parentTypeBuilder the stored record type that other constituents will be extracted from
+     * @see RecordMetaDataBuilder#addUnnestedRecordType(String, String)
+     */
+    @API(API.Status.INTERNAL)
+    public UnnestedRecordTypeBuilder(@Nonnull final String name, @Nonnull Object recordTypeKey, @Nonnull final RecordMetaDataBuilder metaDataBuilder, @Nonnull RecordTypeBuilder parentTypeBuilder) {
+        super(name, recordTypeKey, metaDataBuilder);
+        this.parentTypeBuilder = parentTypeBuilder;
+        this.nestings = new ArrayList<>();
+        // Parent constituent is always present
+        addConstituent(UnnestedRecordType.PARENT_CONSTITUENT, parentTypeBuilder);
+    }
+
+    @API(API.Status.INTERNAL)
+    public UnnestedRecordTypeBuilder(@Nonnull RecordMetaDataProto.UnnestedRecordType typeProto, @Nonnull final RecordMetaDataBuilder metaDataBuilder) {
+        super(typeProto.getName(), LiteralKeyExpression.fromProtoValue(typeProto.getRecordTypeKey()), metaDataBuilder);
+
+        // Parent is a bit special. Go through the constituents and attempt to find it
+        String parentTypeName = null;
+        for (RecordMetaDataProto.UnnestedRecordType.UnnestedConstituent constituentProto : typeProto.getUnnestedConstituentsList()) {
+            if (constituentProto.getName().isEmpty()) {
```

Comment:
Instead of empty string, can we use another constant? I tried to use empty string, but having an empty fieldName in keyExpression seemed to break the index validator. (I'm not sure which line of code it broke, but I tried "__parent" and it worked.)